src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceSinkConfig.java

```diff
@@ -158,21 +207,58 @@ private void validateSchema(Schema schema) {
       throw new InvalidStageException("Sink schema must contain at least one field");
     }
 
-    if (!canAttemptToEstablishConnection() || containsMacro(PROPERTY_SOBJECT)) {
+    if (!canAttemptToEstablishConnection() || containsMacro(PROPERTY_SOBJECT)
+      || containsMacro(PROPERTY_OPERATION) || containsMacro(PROPERTY_EXTERNAL_ID_FIELD)) {
       return;
     }
 
-    Set<String> creatableSObjectFields = null;
-    try {
-      creatableSObjectFields = getCreatableSObjectFields();
-    } catch (ConnectionException e) {
-      throw new InvalidStageException("There was issue communicating with Salesforce", e);
-    }
+    SObjectsDescribeResult describeResult = getSObjectDescribeResult();
+    Set<String> creatableSObjectFields = getCreatableSObjectFields(describeResult);
 
     Set<String> inputFields = schema.getFields()
       .stream()
       .map(Schema.Field::getName)
       .collect(Collectors.toSet());
+
+    OperationEnum operation = getOperationEnum();
+
+    String externalIdFieldName = null;
+    switch (operation) {
+      case insert:
+        break;
+      case upsert:
+        externalIdFieldName = getExternalIdField();
+        break;
+      case update:
+        externalIdFieldName = SALESFORCE_ID_FIELD;
+        break;
+      default:
+        throw new InvalidConfigPropertyException("Unsupported value for operation: " + operation,
+                                                 SalesforceSinkConfig.PROPERTY_OPERATION);
+    }
+
+    if (operation == OperationEnum.upsert) {
+      Field externalIdField = describeResult.getField(sObject, externalIdFieldName);
+      if (externalIdField == null) {
```

Comment:
Actually I just realized that we don't check input schema for types at all. I have added a general check for all the types. This will include checking externalIdField as well.